node/_fs/_fs_read.ts

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+import { Buffer } from "../buffer.ts";
+import { assert } from "../../testing/asserts.ts";
+import { ERR_INVALID_ARG_VALUE } from "../_errors.ts";
+
+type readOptions = {
+  buffer: Buffer | Uint8Array;
+  offset: number;
+  length: number;
+  position: number | null;
+};
+
+type BinaryCallback = (
+  err: Error | null,
+  bytesRead: number | null,
+  data?: Buffer,
+) => void;
+type Callback = BinaryCallback;
+
+export function read(fd: number, callback: Callback): void;
+export function read(
+  fd: number,
+  options: readOptions,
+  callback: Callback,
+): void;
+export function read(
+  fd: number,
+  buffer: Buffer | Uint8Array,
+  offset: number,
+  length: number,
+  position: number | null,
+  callback: Callback,
+): void;
+export function read(
+  fd: number,
+  optOrBuffer?: Buffer | Uint8Array | readOptions | Callback,
+  offsetOrCallback?: number | Callback,
+  length?: number,
+  position?: number | null,
+  callback?: Callback,
+): void {
+  let cb: Callback | undefined;
+  let offset: number = 0,
+    buffer: Buffer | Uint8Array;
+
+  if (length == null) {
+    length = 0;
+  }
+
+  if (typeof offsetOrCallback === "function") {
+    cb = offsetOrCallback;
+  } else if (typeof optOrBuffer === "undefined") {
+    cb = optOrBuffer;
+  } else {
+    offset = offsetOrCallback as number;
+    cb = callback;
+  }
+
+  if (optOrBuffer instanceof Buffer || optOrBuffer instanceof Uint8Array) {
+    buffer = optOrBuffer;
+  } else if (typeof optOrBuffer === "function") {
+    offset = 0;
+    buffer = Buffer.alloc(16384);
+    length = buffer.byteLength;
+    position = null;
+  } else {
+    let opt = optOrBuffer as readOptions;
+    offset = opt.offset;
+    buffer = opt.buffer;
+    position = opt.position;
+    length = opt.length;
```

Comment:
these optional values should have default values such as `opt.offset ?? 0`, `opt.buffer ?? Buffer.alloc(16384)`, etc